### PR TITLE
feat: improve stop timing in drt (towards prebooking)

### PR DESCRIPTION
**Pull request by sebhoerl**
**TL;DR**

- There are incoherences in the timing of stops in DRT (1) when implementing full prebooking with insertion and scheduling, (2) in the treatment of duration changes using MOIA's `(Incremental)StopDurationEstimator`, (3) in the standard case if we don't like requests with zero wait times.
- The PR proposes a new structure for timing with a `StopTimeCalculator` interface that allows to replicate the current stop calculation logic of DRT and to flexibly implement other approaches (cumulative, sequential, prebooking, ...)
- While no applicative changes are made, this is a preparatory PR that will allow to much more easily improve the code for a proper functioning of custom stop durations and prebooking.
- All functional and integration tests pass and replicate current behavior
- Sorry for the long PR, it required quite a bit of refactoring and moving things back and forth. I think this is now a rather boiled down PR to enable all the other planned changes (which are less impacting and more modular).

**Context**

We are currently implementing prebooking in DRT. This means that there are a couple of changes that we'd like to push in the code base, and I'll try to do it step by step. Initially, the problem that arise was that with prebooking, the definition of "stop duration" in DRT becomes ambiguous. If we allow agents to prebook their trips, for instance at the beginning of the day, we may have a request departing at 09:20:20 and another one at 09:20:40, and we can merge them into the same vehicle. However, when should the stop end, when should the pick-ups happen? At the first pickup + 60s or the second pickup + 60s? Or even with individual pickup times during the task (this is the way we are going in our general implementation)? To be congruent with the current functioning of DRT, we came up with the following logic:

- After starting to mount the vehicle, every request needs to pass a finite enter time until it is picked up
- Requests mount in parallel
- A vehicle with assigned requests departs once all pickup requests are on board

This is congruent with the current logic for immediate requests: One or more pick-ups are assigned to a stop. The vehicle arrives, and the stop starts. At this point the agents enter in parallel, and after `stopDuration` the pick-up events are created. While implementing this, I then discovered a couple of inconsistencies already in the existing code base.

**Problem**

- (1) Currently, the insertion algorithm is able to find insertions into *ongoing* (*started*) stop tasks (which are always the first of possible insertion points along a vehicle schedule). The induced delay of adding a new pick-up is zero, the inserter basically bases its calculation on the stop begin time and then adds `stopDuration` which effectively results in an unchanged stop end time. The same logic is then applied in the `RequestInsertionScheduler`. This has the effect that we have requests that are inserted into a stop that already runs for 40s, leading to the request being picked up 20s after departure of the agent. The resulting wait time is hence 20s, which is smaller than the `stopDuration`. Furthermore, we often produce requests with zero wait time. 
- (2) In the current case, one can argue that this is acceptable. With prebooking, the situation will become incoherent. If we are at 05:00 in the morning, and we schedule one request for 09:20:00 and one for 09:20:20 why should one of them wait less, just because the request was submitted a little later? In this case, we should clearly have a (vehicle) stop duration of 80s, which means that we *extend* an existing task.
- (3) The problem already shows up when using the custom `StopDurationEstimator` and `IncrementalStopDurationEstimator` to define custom stop durations. Imagine a wheelchair request that induces an increased stop duration. For any future (planned) stop, adding the request will accordingly extend the stop duration. Likewise, the insertion algorithm will already take into account the extension of the stop when the two interfaces are defined accordingly. Also, `RequestInsertionScheduler` is aware of the interfaces and will adjust the times. However, for an *ongoing* stop, updating the end time of the `DrtStopTask` will not make the `DrtStopActivity` in the QSim finish later. It will just end when it was planned initially.
- (4) Currently, `StopDurationEstimator` and `IncrementalStopDurationEstimator` are kind of awkward to use because the same thing needs to be implemented in different ways at two different spots. In principle, the latter one could just be used to construct the former one.

In summary, in standard DRT, stops are never extended (but maybe they should to avoid zero wait times), and extending stops does not work properly when using a custom `StopDurationEstimator`. Once we add prebooking, just having custom "stop durations" is not enough, but we need to be able to more flexibly define stop end times.

**Solution**

- This PR does not propose a direct solution to the problems above, but it is rather a preparatory step.
- The core is to replace `(Incremental)StopDurationEstimator` with `StopTimeCalculator` which rather operates on stop end times rather than durations (which are tricky, because in prebooking they will depend on `earliestDepartureTime` of the requests in a complex way).
- `StopDurationEstimator` lets users (1) calculate the end time for a new stop with a pickup, (2) update the end time of an existing stop when adding a pickup, (3) calculate the end time for a new stop with a dropoff, (4) update the end time of an existing stop when adding a dropoff, (5) calculating the end time for a stop when the start time is shifted.
- Note that especially (4) is usually the trickiest, because in insertion it does not only depend on the stop that is added, but also on how far the whole vehicle schedule has been shifted from inserting the pickup. This will especially become important for prebooking, where simply shifting the task may *not* actually change its end time if there is enough of "space" before the earliest `earliestDepartureTime` of the requests.
- To give a glimpse on how the implementations will look like, there are the following examples:
  - *DefaultStopTimeCalculator* is the currently active one that reproduces standard DRT behaviour. 
  - *CorrectedStopTimeCalculator* is an update of it that will take into account extending stops when inserting additional pickups to an ongoing stop. *I would like to make this the new standard*, leading all requests to have a wait time of *at least* `stopDuration`. Note that this requires also making `DrtStopTask` responsive to end time changes, which I would solve in a follow-up PR.
  - *ParallelStopTimeCalculator* is a more flexible version of *CorrectedStopTimeCalculator* which operates not on a static `stopDuration` but obtains the duration from a flexible `StopDurationProvider`. It assumes that agents enter in parallel, so the updating the end time of the stop is roughly `max(currentStopEndTime, currentStopBeginTime + customStopDuration`.
  - *CumulativeStopTimeCalculator* is a different implementation that makes use of a `StopDurationProvider` and resembles the logic that has been defined in MOIA's integration tests for defining flexible stop durations. Basically, when having `N` pickups in the task, the end time will be calculated based on the sum of the individual stop times.
  - *PrebookingStopTimeCalculator* is a preview for our prebooking implementation, showing the increased complexity when `earliestDepartureTime`s of all assigned requests need to be considered.

**Roadmap for next PRs**

- Properly handling the extension of stops 
  - Make `DrtStopActivity` depend dynamically on the stop end time
  - There is a caveat: Extending a stop may move the end time (= pickup time) of a stop beyond the `latestDepartureTime` of a request. This should then not be a valid insertion. Currently, the *slack* calculations in the insertion algorithm only denote how much space for insertion we have *after* an insertion index. To cover for the new edge case, we should calculate some *start slack* that indicates how far we can extend the first stop task.
  - This then will allow to (maybe optionally) avoid zero-wait-time requests and fix variable stop durations.
- Prebooking
  - Will allow the user to submit requests at any time of the day for a later departure time
  - Also requires a more dynamic version of `DrtStopActivity`
  - We have a working version internally, I'll slice it up in digestable PRs :)

### CODE REVIEW

File: contribs/drt/src/main/java/org/matsim/contrib/drt/optimizer/insertion/InsertionDetourTimeCalculator.java [hunk 1]
```diff
@@ -157,8 +154,33 @@ private PickupTimeInfo calculatePickupIfSameLink(VehicleEntry vEntry, int pickup
...
+			 * TODO: Generally, DRT is allowed to insert pickups into ongoing stop tasks.
+			 * However, currently, stop tasks have a fixed duration that is never changed.
+			 * This means that a pickup is added to an ongoing task that will end, maybe in
+			 * 20s. The task will then end regardless of the added pickup. The request will,
+			 * hence, have a wait time of 20s although the configured stop duration may be
+			 * 60s. We can even frequently have requests with zero wait time.
+			 * 
+			 * To mitigate the problem, the following steps are necessary:
+			 * - Impose here not the stopTask.beginTime as the time at which the request is inserted,
+			 *   but impose the current time ("now") as the point of insertion.
```
Comment (michalmac) [inline contribs/drt/src/main/java/org/matsim/contrib/drt/optimizer/insertion/InsertionDetourTimeCalculator.java:167]: We could provide both: (1) the start of the task and (2) the current time. This way we can also get the old behaviour.

I am not saying that the old behaviour is best. To be honest, I would like to have a minimum viable flexibility, so the optimizer code is as simple as possible. I could also imagine that at some point it's easier to have two separate insertion algorithms, each supporting a subset of features than to have one super-flexible that is hard to understand, maintain and fix.

**Review by michalmac — COMMENTED** [review]
This is a nice cleanup wrt setting/updating start/end times of stop tasks.

A few general comments:
- Could you update the commit messages so they are less cryptic than "works" or "update".
- As always I am a bit afraid of being to generic/flexible at the cost of adding more layers and abstractions, because DRT is already too complex (therefore I suggested a removal of one interface). I added a few more related nodes in separate code comments.

Let's maybe have one more round of discussion before merging this PR, but in general I think we should merge it.

File: contribs/drt/src/main/java/org/matsim/contrib/drt/optimizer/insertion/InsertionDetourTimeCalculator.java (hunk 1 above)
Comment (sebhoerl) [reply contribs/drt/src/main/java/org/matsim/contrib/drt/optimizer/insertion/InsertionDetourTimeCalculator.java:167]: We pass the DrtStopTask, which contains the stop begin time. So we can exactly replicate the existing behavior, which is also the stanard implementation. So in the standard setting, this is only another parameter that is pasesd to the StopTimeCalculator, but not used. However, for more complex implementation it can be used. In general the idea was to touch the insertion code as little as possible in this PR and extract the stop time logic into the StopTimeCalculator.

File: contribs/drt/src/main/java/org/matsim/contrib/drt/run/DrtModeModule.java [hunk 2]
```diff
@@ -81,10 +83,11 @@ public void install() {
 							getter.getModal(DrtEventSequenceCollector.class)))).asEagerSingleton();
 			addControlerListenerBinding().to(modalKey(DrtSpeedUp.class));
 		});
-
-		bindModal(StopDurationEstimator.class).toInstance(
-				(vehicle, dropoffRequests, pickupRequests) -> drtCfg.stopDuration);
-		bindModal(IncrementalStopDurationEstimator.class).toInstance(
-				new DefaultIncrementalStopDurationEstimator(drtCfg.stopDuration));
+		
+		bindModal(DefaultStopTimeCalculator.class).toProvider(modalProvider(getter -> {
```
[the remark this answers is not in the document]
Comment (sebhoerl) [reply contribs/drt/src/main/java/org/matsim/contrib/drt/run/DrtModeModule.java:86]: For me both are fine, I can change it if you prefer. I like to have a specific binding for the default implementation so it becomes easier to write decorators and inject the default implementation.

File: contribs/drt/src/main/java/org/matsim/contrib/drt/run/DrtModeModule.java [hunk 3]
```diff
@@ -23,11 +23,12 @@
 import org.matsim.api.core.v01.network.Network;
 import org.matsim.contrib.drt.analysis.DrtEventSequenceCollector;
 import org.matsim.contrib.drt.fare.DrtFareHandler;
-import org.matsim.contrib.drt.optimizer.insertion.DefaultIncrementalStopDurationEstimator;
-import org.matsim.contrib.drt.optimizer.insertion.IncrementalStopDurationEstimator;
 import org.matsim.contrib.drt.optimizer.rebalancing.RebalancingModule;
-import org.matsim.contrib.drt.schedule.StopDurationEstimator;
 import org.matsim.contrib.drt.speedup.DrtSpeedUp;
+import org.matsim.contrib.drt.stops.DefaultStopDurationProvider;
```
[the remark this answers is not in the document]
Comment (sebhoerl) [reply contribs/drt/src/main/java/org/matsim/contrib/drt/run/DrtModeModule.java]: I think this refers to an earlier commit, doesn't exist anymore in the current version.

File: contribs/drt/src/main/java/org/matsim/contrib/drt/optimizer/insertion/InsertionDetourTimeCalculator.java [hunk 4]
```diff
@@ -35,175 +36,152 @@ public InsertionDetourTimeCalculator(StopTimeCalculator stopTimeCalculator,
...
-			double pickupTimeLoss = additionalStopDuration + fromPickupTT - replacedDriveTT;
-			
-			return new PickupDetourInfo(departureTime, pickupTimeLoss);
-		} else {
-			// no preceding stop task, but toPickupTT may be zero if it is the same link (case 1)
-			double arrivalTime = pickup.previousWaypoint.getDepartureTime() + toPickupTT;
+												 double fromPickupTT, boolean followedByDropoff, DrtRequest drtRequest) {
+		if (followedByDropoff) {
+			//handle the pickup->dropoff case separately
+			return calcPickupDetourInfoIfPickupToDropoffDetour(vEntry, pickup, toPickupTT, fromPickupTT, drtRequest);
```
[the remark this answers is not in the document]
Comment (sebhoerl) [reply contribs/drt/src/main/java/org/matsim/contrib/drt/optimizer/insertion/InsertionDetourTimeCalculator.java]: I think this refers to an earlier commit, doesn't exist anymore in the current version of `InsertionDetourTimeCalculator`.

**Comment by sebhoerl** [comment]
Hi @michalmac, I responded to your comments. Also, #2730 may add some explanation.

About your general comments:
- Some of the commits were work-in-progress, I was assuming we squash everything and have the final version as one commit. I can do a rebase already if you prefer.
- In that spirit I tried to limit the changes in this PR and continue to tackle the specific timing fixes in  #2730.
- imo this PR does not really add complexity, it just replaces and existing interface (providing stop duration) with a new one (providing stop end time) that allows for more (optional) complexity
- the additional `StopDurationProvider` is an implementation-specific interface (behind certain `StopTimeCalculator` implementations) that does not directly interfere with the drt insertion/scheduling code

What do you think?

File: contribs/drt-extensions/src/main/java/org/matsim/contrib/drt/extension/operations/shifts/optimizer/ShiftRequestInsertionScheduler.java [hunk 5]
```diff
@@ -220,9 +225,9 @@ private DrtStopTask insertPickup(AcceptedDrtRequest request, InsertionWithDetour
         // insert pickup stop task
         double startTime = beforePickupTask.getEndTime();
         int taskIdx = beforePickupTask.getTaskIdx() + 1;
-		double stopDuration = stopDurationEstimator.calcDuration(vehicleEntry.vehicle, Collections.emptySet(), Collections.singleton(request));
-        DrtStopTask pickupStopTask = taskFactory.createStopTask(vehicleEntry.vehicle, startTime,
-                Math.max(startTime + stopDuration, request.getEarliestStartTime()), request.getFromLink());
+        double stopEndTime = stopTimeCalculator.initEndTimeForPickup(vehicleEntry.vehicle, startTime, request.getRequest());
```
[the remark this answers is not in the document]
Comment (michalmac) [reply contribs/drt-extensions/src/main/java/org/matsim/contrib/drt/extension/operations/shifts/optimizer/ShiftRequestInsertionScheduler.java:228]: This and a few other of my comments refer to specific commits, because I often review commit by commit.

File: contribs/drt/src/main/java/org/matsim/contrib/drt/optimizer/insertion/InsertionDetourTimeCalculator.java (hunk 1 above)
Comment (michalmac) [reply contribs/drt/src/main/java/org/matsim/contrib/drt/optimizer/insertion/InsertionDetourTimeCalculator.java:167]: I understood your comment that you want to change the current behaviour. Now I got it. Thanks.

File: contribs/drt/src/main/java/org/matsim/contrib/drt/run/DrtModeModule.java (hunk 2 above)
[the remark this answers is not in the document]
Comment (michalmac) [reply contribs/drt/src/main/java/org/matsim/contrib/drt/run/DrtModeModule.java:86]: I guess this is OK. We do not have any "code style for guice bindings", so both approaches are equally good. Just wanted to know your opinion.

File: contribs/drt/src/main/java/org/matsim/contrib/drt/stops/StopDurationProvider.java [hunk 6]
```diff
@@ -0,0 +1,15 @@
...
+
+import org.matsim.contrib.drt.passenger.DrtRequest;
+import org.matsim.contrib.dvrp.fleet.DvrpVehicle;
+
+/**
+ * This class provides the stop duration of a request. Eventually, this could
+ * also become a simple attribute of a DrtRequest to simplify the code
+ * complexity.
+ */
+public interface StopDurationProvider {
```
[the remark this answers is not in the document]
Comment (michalmac) [reply contribs/drt/src/main/java/org/matsim/contrib/drt/stops/StopDurationProvider.java:11]: If we assume the stop duration is a linear function (a fixed duration + sum of stop duration of each request), we could call it `StopDurationIncreaseCalculator` or `PassengerStopDurationCalculator` or, following your suggestion, `StopInteractionDurationCalculator`. And accordingly change the method names. Perhaps the last one is best if we want to use some non-linear functions in the future.

I would rather not put it into the request (as a standard attribute), because it is very hard to later remove/modify it.

File: contribs/drt/src/main/java/org/matsim/contrib/drt/optimizer/insertion/InsertionDetourTimeCalculator.java [hunk 7]
```diff
@@ -157,8 +154,33 @@ private PickupTimeInfo calculatePickupIfSameLink(VehicleEntry vEntry, int pickup
...
 			
-			// IRTX TODO: This should actually become the *actual* time of insertion which
-			// may be *now*, i.e. a couple of seconds after the task has been started
+			/*
+			 * TODO: Generally, DRT is allowed to insert pickups into ongoing stop tasks.
+			 * However, currently, stop tasks have a fixed duration that is never changed.
+			 * This means that a pickup is added to an ongoing task that will end, maybe in
+			 * 20s. The task will then end regardless of the added pickup. The request will,
+			 * hence, have a wait time of 20s although the configured stop duration may be
+			 * 60s. We can even frequently have requests with zero wait time.
```
[the remark this answers is not in the document]
Comment (michalmac) [reply contribs/drt/src/main/java/org/matsim/contrib/drt/optimizer/insertion/InsertionDetourTimeCalculator.java:163]: Makes sense. I just wanted to mention that 

>We can even frequently have requests with zero wait time.

is not a bug given the current assumptions. I understand that different problems/simulations may require different assumptions.

File: contribs/drt/src/main/java/org/matsim/contrib/drt/stops/StopDurationProvider.java (hunk 6 above)
[the remark this answers is not in the document]
Comment (sebhoerl) [reply contribs/drt/src/main/java/org/matsim/contrib/drt/stops/StopDurationProvider.java:11]: So taking the suggestions, I would propose `PassengerStopDurationProvider`. I would like to call it "provider" to convey that we really just want to obtain a value here, the arbitrarily complex calculation should happen in the specific `StopTimeCalculator`.

File: contribs/drt-extensions/src/main/java/org/matsim/contrib/drt/extension/operations/shifts/optimizer/ShiftRequestInsertionScheduler.java (hunk 5 above)
[the remark this answers is not in the document]
Comment (sebhoerl) [reply contribs/drt-extensions/src/main/java/org/matsim/contrib/drt/extension/operations/shifts/optimizer/ShiftRequestInsertionScheduler.java:228]: Ok, i see. in the current version the drt-extensions tests should pass

File: contribs/drt/src/main/java/org/matsim/contrib/drt/optimizer/insertion/InsertionDetourTimeCalculator.java (hunk 7 above)
[the remark this answers is not in the document]
Comment (sebhoerl) [reply contribs/drt/src/main/java/org/matsim/contrib/drt/optimizer/insertion/InsertionDetourTimeCalculator.java:163]: Ok, I would propose then to stick everywhere to the current behavior, also in #2370 and later start a discussion if any defaults should be changed or not and how to make this transparent

**Comment by michalmac** [comment]
> Hi @michalmac, thanks for your review, I responded to your comments. Also, #2730 may add some explanation.

Thanks for your replies!

> 
> About your general comments:
> 
> * Some of the commits were work-in-progress, I was assuming we squash everything and have the final version as one commit. I can do a rebase already if you prefer.

Squashing is fine.

> * In that spirit I tried to limit the changes in this PR and continue to tackle the specific timing fixes in  [feat: avoid short wait times in drt #2730](https://github.com/matsim-org/matsim-libs/pull/2730).

I haven't yet got there :-)

> * imo this PR does not really add complexity, it just replaces and existing interface (providing stop duration) with a new one (providing stop end time) that allows for more (optional) complexity

Yes, I would say the complexity was already added some time ago with the first interface, while this PR refines the interfaces, which is nice.

To help the users and developers, maybe we should reduce ambiguities around:
- `StopDurationProvider` (by renaming it?)
- `stopDuration` in the drt config (To me, this is the minimum stop duration regardless of the number of passengers picked up/dropped off; not a stop duration per passenger; not a minimum wait time. Maybe we should update the comment for that field?)

> * the additional `StopDurationProvider` is an implementation-specific interface (behind certain `StopTimeCalculator` implementations) that does not directly interfere with the drt insertion/scheduling code

Here's my related comment: https://github.com/matsim-org/matsim-libs/pull/2723#discussion_r1295795581

> What do you think?

I think it looks good. Before merging, maybe we should try to reduce the two ambiguities I mentioned above?

File: contribs/drt/src/main/java/org/matsim/contrib/drt/optimizer/insertion/InsertionDetourTimeCalculator.java (hunk 7 above)
[the remark this answers is not in the document]
Comment (michalmac) [reply contribs/drt/src/main/java/org/matsim/contrib/drt/optimizer/insertion/InsertionDetourTimeCalculator.java:163]: Sure, we can change the defaults.

> also in #2370 and later

 I guess you meant #2730 :-)

**Review by michalmac — APPROVED** [review]
Nice work! ⭐
